File: wizard.py
"""Interactive wizard steps — collects input file, preset, and output location."""

from pathlib import Path

import config

HERE = Path(__file__).parent


def ask(prompt: str, default: str = "") -> str:
    hint = f" (default: {default})" if default else ""
    val = input(f"  {prompt}{hint}:\n  > ").strip().strip('"').strip("'")
    return val if val else default


def divider():
    print("\n" + "─" * 44 + "\n")


def _clean_path(raw: str) -> str:
    raw = raw.strip()
    if raw.startswith("& "):   # PowerShell drag-and-drop prefix
        raw = raw[2:]
    return raw.strip('"').strip("'")


def step_input() -> Path:
    print("STEP 1 - Input file")
    print("  Drag & drop a file into this window, or paste the full path.\n")
    while True:
        path = Path(_clean_path(input("  > ")))
        if path.exists() and path.is_file():
            return path
        print(f"\n  [!] Can't find that file — try again.\n")


def step_preset() -> tuple[int | None, list[str], str]:
    """Returns (size_limit_bytes or None, codec_priority, label)."""
    print("STEP 2 — Target preset")
    print("  What are you compressing for?\n")

    preset_list = config.presets()
    for p in preset_list:
        print(f"  [{p['key']}]  {p['label']}")
    print()

    preset_map = {p["key"]: p for p in preset_list}
    while True:
        choice = input("  > ").strip()
        if choice in preset_map:
            p = preset_map[choice]
            break
        print(f"  [!] Enter a number 1–{len(preset_list)}.\n")

    size_limit     = int(p["size_mb"] * 1024 * 1024) if p["size_mb"] else None
    codec_priority = list(p["codecs"])
    label          = p["label"]

    if p["id"] == "custom":
        print()
        while True:
            raw = input("  Custom size limit in MB (or leave blank for none):\n  > ").strip()
            if not raw:
                size_limit = None
                break
            try:
                size_limit = int(float(raw) * 1024 * 1024)
                label = f"Custom ({raw} MB)"
                break
            except ValueError:
                print("  [!] Enter a number like 50\n")

    return size_limit, codec_priority, label


def step_encoder(working: list[dict], recommended: str | None) -> str | None:
    """Show working encoders with descriptions; let the user pick one."""
    print("STEP 3 - Encoder")
    print("  Choose which encoder to use. Press Enter for the recommended option.\n")

    if not working:
        return None

    for i, enc in enumerate(working, 1):
        rec  = "  <- recommended" if enc["name"] == recommended else ""
        slow = "  [!] very slow"  if enc["slow"] else ""
        print(f"  [{i}]  {enc['label']}{rec}{slow}")
        print(f"       {enc['desc']}")
        print()

    default_i = next((i for i, e in enumerate(working, 1) if e["name"] == recommended), 1)

    while True:
        raw = input(f"  > (default: {default_i})\n  > ").strip()
        if not raw:
            return working[default_i - 1]["name"]
        if raw.isdigit():
            idx = int(raw)
            if 1 <= idx <= len(working):
                chosen = working[idx - 1]
                if chosen["slow"]:
                    print()
                    print("  [!] This encoder is extremely slow.")
                    print("      A 10-minute video can easily take several hours.")
                    print()
                    if input("  Are you sure? [y/N]: ").strip().lower() != "y":
                        print()
                        continue
                return chosen["name"]
        print(f"  [!] Enter a number 1-{len(working)}, or press Enter for the default.\n")


def step_resolution(src_height: int, complexity: str, steps: list[int]) -> int:
    """
    Returns 0 for Auto, or a forced output height (int).
    If the returned height >= src_height, no scaling is applied.
    """
    print("STEP 4 - Resolution")
    print("  Choose an output resolution, or Auto to let the tool decide.\n")

    res_info = config.resolution_info()

    # Build option list: Auto (0) + source height + all standard steps below source
    options: list[int] = [0]
    if src_height not in steps:
        options.append(src_height)
    for h in steps:
        if h <= src_height:
            options.append(h)

    # Recommendation: for Complex content at 1080p+, suggest one step down
    if complexity == "Complex" and src_height >= 1080:
        recommended = next((h for h in steps if h < src_height), 0)
    else:
        recommended = 0  # Auto

    for i, res in enumerate(options, 1):
        if res == 0:
            label = "Auto"
            desc  = res_info.get("auto", "")
        elif res == src_height:
            label = f"{res}p (original)"
            desc  = res_info.get(str(res), "") or "Your source resolution."
        else:
            label = f"{res}p"
            desc  = res_info.get(str(res), "")
        rec = "  <- recommended" if res == recommended else ""
        print(f"  [{i}]  {label}{rec}")
        print(f"       {desc}")
        print()

    default_i = next((i for i, r in enumerate(options, 1) if r == recommended), 1)

    while True:
        raw = input(f"  > (default: {default_i})\n  > ").strip()
        if not raw:
            return options[default_i - 1]
        if raw.isdigit():
            idx = int(raw)
            if 1 <= idx <= len(options):
                return options[idx - 1]
        print(f"  [!] Enter a number 1-{len(options)}, or press Enter for the default.\n")


def step_output(input_path: Path) -> tuple[str, Path]:
    """Returns (filename, output_folder)."""
    print("STEP 5 - Output filename")
    default_name = f"{input_path.stem}_compressed.mp4"
    out_name = ask("Filename", default_name)
    if not out_name.lower().endswith(".mp4"):
        out_name += ".mp4"

    divider()

    print("STEP 6 - Output folder")
    default_folder = str(HERE / "out")
    raw_folder = ask("Folder", default_folder)
    out_folder = Path(_clean_path(raw_folder))
    out_folder.mkdir(parents=True, exist_ok=True)

    # Warn if output already exists
    out_path = out_folder / out_name
    while out_path.exists():
        print(f"\n  [!] '{out_name}' already exists in that folder.")
        print("  [1]  Overwrite it")
        print("  [2]  Choose a different name")
        print()
        choice = input("  > ").strip()
        if choice == "1":
            break
        elif choice == "2":
            base = Path(out_name).stem.removesuffix("_compressed")
            counter = 2
            suggested = f"{base}_compressed_{counter}.mp4"
            while (out_folder / suggested).exists():
                counter += 1
                suggested = f"{base}_compressed_{counter}.mp4"
            out_name = ask("New filename", suggested)
            if not out_name.lower().endswith(".mp4"):
                out_name += ".mp4"
            out_path = out_folder / out_name
        else:
            print("  [!] Enter 1 or 2.\n")

    return out_name, out_folder

File: SmartCompress.CLI/Wizard.cs
using System.Globalization;
using SmartCompress.Core;
using Spectre.Console;

namespace SmartCompress;

internal static class Wizard
{
    public static void Divider() =>
        AnsiConsole.Write(new Rule().RuleStyle("grey dim"));

    /// <summary>
    /// `Console.ReadLine` that first drains any buffered keystrokes the user
    /// typed during the previous encode. Prevents accidentally skipping an
    /// important prompt because they hit Enter while a progress bar was up.
    /// Returns the empty string if input is closed (Ctrl+Z / piped EOF).
    /// </summary>
    public static string ReadLineDrained()
    {
        if (!Console.IsInputRedirected)
        {
            try
            {
                while (Console.KeyAvailable)
                    Console.ReadKey(intercept: true);
            }
            catch { /* terminal can't peek — skip */ }
        }
        return Console.ReadLine() ?? "";
    }

    private static string Ask(string prompt, string @default = "")
    {
        var hint = @default.Length > 0 ? $" [dim](default: {Markup.Escape(@default)})[/]" : "";
        AnsiConsole.MarkupLine($"  {Markup.Escape(prompt)}{hint}:");
        AnsiConsole.Markup("  [dim]>[/] ");
        var val = ReadLineDrained().Trim().Trim('"').Trim('\'');
        return val.Length > 0 ? val : @default;
    }

    private static string CleanPath(string raw)
    {
        raw = raw.Trim();
        if (raw.StartsWith("& ")) raw = raw[2..];
        return raw.Trim('"').Trim('\'');
    }

    // ── Step 1: Input file or folder ─────────────────────────────────────────

    /// <summary>
    /// Accepts either a single file OR a directory. The dispatcher figures
    /// out later which collected files are actually supported media.
    /// Returns the original raw path (caller passes it through `BatchCollector`).
    /// </summary>
    public static string StepInput()
    {
        AnsiConsole.MarkupLine("[bold cyan]STEP 1[/] [bold]— Input[/]");
        AnsiConsole.MarkupLine("  [dim]Drag & drop a file OR folder into this window, or paste the path.[/]");
        AnsiConsole.MarkupLine("  [dim]Folders are scanned (up to 2 levels deep) and unsupported files are skipped.[/]\n");
        while (true)
        {
            AnsiConsole.Markup("  [dim]>[/] ");
            var path = CleanPath(ReadLineDrained());
            if (File.Exists(path) || Directory.Exists(path)) return path;
            AnsiConsole.MarkupLine("\n  [red][[!]] Can't find that file or folder — try again.[/]\n");
        }
    }

    // ── Step 2: Preset ────────────────────────────────────────────────────────

    public static (long? SizeLimit, List<string> CodecPriority, string PresetLabel) StepPreset(AppConfig config)
    {
        AnsiConsole.MarkupLine("[bold cyan]STEP 2[/] [bold]— Target preset[/]");
        AnsiConsole.MarkupLine("  [dim]What are you compressing for?[/]\n");

        var presets = config.Presets;
        foreach (var p in presets)
            AnsiConsole.MarkupLine($"  [dim][[{p.Key}]][/]  {Markup.Escape(p.Label)}");
        Console.WriteLine();

        var presetMap = presets.ToDictionary(p => p.Key);
        Preset chosen = default!;
        while (true)
        {
            AnsiConsole.Markup("  [dim]>[/] ");
            var key = ReadLineDrained().Trim();
            if (presetMap.TryGetValue(key, out chosen!)) break;
            AnsiConsole.MarkupLine($"  [red][[!]] Enter a number 1–{presets.Count}.[/]\n");
        }

        long? sizeLimit = chosen.SizeMb.HasValue ? (long)(chosen.SizeMb.Value * 1_048_576) : null;
        var codecs = new List<string>(chosen.Codecs);
        var label  = chosen.Label;

        if (chosen.Id == "custom")
        {
            Console.WriteLine();
            while (true)
            {
                AnsiConsole.MarkupLine("  [dim]Custom size limit in MB (or leave blank for none):[/]");
                AnsiConsole.Markup("  [dim]>[/] ");
                var raw = ReadLineDrained().Trim();
                if (raw.Length == 0) { sizeLimit = null; break; }
                if (double.TryParse(raw, NumberStyles.Float, CultureInfo.InvariantCulture, out var mb))
                {
                    sizeLimit = (long)(mb * 1_048_576);
                    label     = $"Custom ({raw} MB)";
                    break;
                }
                AnsiConsole.MarkupLine("  [red][[!]] Enter a number like 50[/]\n");
            }
        }

        return (sizeLimit, codecs, label);
    }

    // ── Step 3: Encoder ───────────────────────────────────────────────────────

    public static string? StepEncoder(List<EncoderEntry> working, string? recommended)
    {
        AnsiConsole.MarkupLine("[bold cyan]STEP 3[/] [bold]— Encoder[/]");
        AnsiConsole.MarkupLine("  [dim]Choose which encoder to use. Press Enter for the recommended option.[/]\n");

        if (working.Count == 0) return null;

        for (int i = 0; i < working.Count; i++)
        {
            var enc      = working[i];
            var isRec    = enc.Name == recommended;
            var labelColor = enc.Speed switch
            {
                EncoderSpeed.VerySlow => "red",
                EncoderSpeed.Slow     => "yellow",
                _                     => "green"
            };
            var slowTag = enc.Speed switch
            {
                EncoderSpeed.VerySlow => "  [bold red][[!]] very slow[/]",
                EncoderSpeed.Slow     => "  [dim yellow][[slow]][/]",
                _                     => ""
            };
            var recTag = isRec ? "  [bold green]← recommended[/]" : "";

            AnsiConsole.MarkupLine(
                $"  [dim][[{i + 1}]][/]  [{labelColor}]{Markup.Escape(enc.Label)}[/]{recTag}{slowTag}");
            AnsiConsole.MarkupLine($"       [dim]{Markup.Escape(enc.Desc)}[/]");
            Console.WriteLine();
        }

        int defaultIdx = working.FindIndex(e => e.Name == recommended);
        if (defaultIdx < 0) defaultIdx = 0;

        while (true)
        {
            AnsiConsole.MarkupLine($"  [dim](default: {defaultIdx + 1})[/]");
            AnsiConsole.Markup("  [dim]>[/] ");
            var raw = ReadLineDrained().Trim();
            if (raw.Length == 0) return working[defaultIdx].Name;

            if (int.TryParse(raw, out var idx) && idx >= 1 && idx <= working.Count)
            {
                var enc = working[idx - 1];
                if (enc.Speed == EncoderSpeed.VerySlow)
                {
                    Console.WriteLine();
                    AnsiConsole.MarkupLine("  [bold red][[!]] This encoder is extremely slow.[/]");
                    AnsiConsole.MarkupLine("      [dim]A 10-minute video can easily take several hours.[/]");
                    Console.WriteLine();
                    AnsiConsole.Markup("  Are you sure? [dim][[y/N]][/]: ");
                    if (ReadLineDrained().Trim().ToLowerInvariant() != "y")
                    { Console.WriteLine(); continue; }
                }
                return enc.Name;
            }
            AnsiConsole.MarkupLine($"  [red][[!]] Enter a number 1-{working.Count}, or press Enter for the default.[/]\n");
        }
    }

    // ── Step 4: Resolution ────────────────────────────────────────────────────

    public static int StepResolution(int srcHeight, string complexity, AppConfig config)
    {
        AnsiConsole.MarkupLine("[bold cyan]STEP 4[/] [bold]— Resolution[/]");
        AnsiConsole.MarkupLine("  [dim]Choose an output resolution, or Auto to let the tool decide.[/]\n");

        var resInfo = config.ResolutionInfo;
        var steps   = config.ResolutionSteps;

        var options = new List<int> { 0 };
        if (!steps.Contains(srcHeight)) options.Add(srcHeight);
        foreach (var h in steps)
            if (h <= srcHeight) options.Add(h);

        int recommended = (complexity == "Complex" && srcHeight >= 1080)
            ? steps.FirstOrDefault(h => h < srcHeight)
            : 0;

        for (int i = 0; i < options.Count; i++)
        {
            int res = options[i];
            string label, desc;
            if (res == 0)        { label = "Auto";                desc = resInfo.GetValueOrDefault("auto", ""); }
            else if (res == srcHeight) { label = $"{res}p (original)"; desc = resInfo.GetValueOrDefault(res.ToString(), "") ?? "Your source resolution."; }
            else                 { label = $"{res}p";             desc = resInfo.GetValueOrDefault(res.ToString(), ""); }

            var recTag = res == recommended ? "  [bold green]← recommended[/]" : "";
            AnsiConsole.MarkupLine($"  [dim][[{i + 1}]][/]  [white]{Markup.Escape(label)}[/]{recTag}");
            AnsiConsole.MarkupLine($"       [dim]{Markup.Escape(desc ?? "")}[/]");
            Console.WriteLine();
        }

        int defaultIdx = options.IndexOf(recommended);
        if (defaultIdx < 0) defaultIdx = 0;

        while (true)
        {
            AnsiConsole.MarkupLine($"  [dim](default: {defaultIdx + 1})[/]");
            AnsiConsole.Markup("  [dim]>[/] ");
            var raw = ReadLineDrained().Trim();
            if (raw.Length == 0) return options[defaultIdx];
            if (int.TryParse(raw, out var idx) && idx >= 1 && idx <= options.Count)
                return options[idx - 1];
            AnsiConsole.MarkupLine($"  [red][[!]] Enter a number 1-{options.Count}, or press Enter for the default.[/]\n");
        }
    }

    // ── Step: Output folder only (batch mode — pipelines pick filenames) ─────

    /// <summary>
    /// Asks for an output folder only. Pipelines derive per-file output names
    /// from each input. Falls back to env var or platform default if blank.
    /// </summary>
    public static string StepOutputFolder()
    {
        AnsiConsole.MarkupLine("[bold cyan]Output folder[/]");
        var scOutDir = Environment.GetEnvironmentVariable("SC_OUT_DIR");
        var defaultFolder = scOutDir is not null
            ? Path.Combine(scOutDir, "out")
            : AppPaths.DefaultOutputDir;
        var rawFolder = Ask("Folder", defaultFolder);
        var outFolder = CleanPath(rawFolder);
        Directory.CreateDirectory(outFolder);
        return outFolder;
    }

    // ── Steps 5 + 6: Output filename and folder (single-file legacy path) ────

    public static (string FileName, string Folder) StepOutput(string inputPath)
    {
        var stem        = Path.GetFileNameWithoutExtension(inputPath);
        var defaultName = $"{stem}_compressed.mp4";

        AnsiConsole.MarkupLine("[bold cyan]STEP 5[/] [bold]— Output filename[/]");
        var outName = Ask("Filename", defaultName);
        if (!outName.EndsWith(".mp4", StringComparison.OrdinalIgnoreCase)) outName += ".mp4";

        Divider();
        Console.WriteLine();

        AnsiConsole.MarkupLine("[bold cyan]STEP 6[/] [bold]— Output folder[/]");
        var scOutDir      = Environment.GetEnvironmentVariable("SC_OUT_DIR");
        var defaultFolder = scOutDir is not null
            ? Path.Combine(scOutDir, "out")
            : AppPaths.DefaultOutputDir;
        var rawFolder = Ask("Folder", defaultFolder);
        var outFolder = CleanPath(rawFolder);
        Directory.CreateDirectory(outFolder);

        var outPath = Path.Combine(outFolder, outName);
        while (File.Exists(outPath))
        {
            AnsiConsole.MarkupLine($"\n  [yellow][[!]] '{Markup.Escape(outName)}' already exists in that folder.[/]");
            AnsiConsole.MarkupLine("  [dim][[1]][/]  Overwrite it");
            AnsiConsole.MarkupLine("  [dim][[2]][/]  Choose a different name");
            Console.WriteLine();
            AnsiConsole.Markup("  [dim]>[/] ");
            var choice = ReadLineDrained().Trim();
            if (choice == "1") break;
            if (choice == "2")
            {
                var baseName = Path.GetFileNameWithoutExtension(outName);
                if (baseName.EndsWith("_compressed", StringComparison.Ordinal)) baseName = baseName[..^11];
                int counter   = 2;
                var suggested = $"{baseName}_compressed_{counter}.mp4";
                while (File.Exists(Path.Combine(outFolder, suggested)))
                    suggested = $"{baseName}_compressed_{++counter}.mp4";
                outName = Ask("New filename", suggested);
                if (!outName.EndsWith(".mp4", StringComparison.OrdinalIgnoreCase)) outName += ".mp4";
                outPath = Path.Combine(outFolder, outName);
            }
            else AnsiConsole.MarkupLine("  [red][[!]] Enter 1 or 2.[/]\n");
        }

        return (outName, outFolder);
    }
}
